tracing-subscriber/src/filter/layer_filters/mod.rs

```diff
@@ -579,7 +579,9 @@ where
 
     fn on_new_span(&self, attrs: &span::Attributes<'_>, id: &span::Id, cx: Context<'_, S>) {
         self.did_enable(|| {
-            self.layer.on_new_span(attrs, id, cx.with_filter(self.id()));
+            let cx = cx.with_filter(self.id());
+            self.filter.on_new_span(attrs, id, cx.clone());
```

Comment:
I mostly guessed here: should the `filter.on_new_span` method be called with `cx.with_filter` or just with `cx`? I don't really understand these details here.

Also, I guessed that it might be more correct to first update the filters and then the layer, not sure if that can make a difference or of the filter should be called last.

Reply:
> I mostly guessed here: should the `filter.on_new_span` method be called with `cx.with_filter` or just with `cx`? I don't really understand these details here.

This is a good question. The difference is that calling `cx.with_filter(...)` means that the context will filter out any spans from the current span context that were previously disabled by that filter. I _think_ it makes sense for the filter to not see any spans that it previously filtered out. So, passing the `with_filter` context to the filter is probably the right thing.

> Also, I guessed that it might be more correct to first update the filters and then the layer, not sure if that can make a difference or of the filter should be called last. The question about the order of `filter` and `layer` calls applies to all four changes in this file

I don't think it actually makes a difference which order the methods are called in. The filter can only enable/disable things in its `enabled` method, not in `on_$WHATEVER`, so calling the `on_$WHATEVER` method is only going to effect the _next `enabled` call_. It won't have an effect on what the inner layer sees in its `on_$WHATEVER` method. So, we can actually call them in whatever order and it shouldn't make a difference.

Reply:
> I think it makes sense for the filter to not see any spans that it previously filtered out.
that sounds to me like it might depend on what `Interest` that filter would return for that span? Very low confidence though 😅 

Good to know about the order though. Thanks for the explanations!